deltacat/compute/compactor_v2/steps/merge.py

```diff
@@ -1,18 +1,439 @@
+            manifest=manifest,
+            previous_stream_position=write_to_partition.stream_position,
+            properties={},
+        )
+        materialize_result = MaterializeResult.of(
+            delta=delta,
+            task_index=hb_index,
+            pyarrow_write_result=None,
+            # TODO (pdames): Generalize WriteResult to contain in-memory-table-type
+            #  and in-memory-table-bytes instead of tight coupling to paBytes
+            referenced_pyarrow_write_result=PyArrowWriteResult.of(
+                len(manifest_entry_referenced_list),
+                manifest.meta.source_content_length,
+                manifest.meta.content_length,
+                manifest.meta.record_count,
+            ),
+        )
+        materialize_result_list.append(materialize_result)
+    return materialize_result_list
+
+
 def _timed_merge(input: MergeInput) -> MergeResult:
-    # TODO: Implementation goes here
-    pass
+    def _materialize(
+        hash_bucket_index,
+        compacted_tables: List[pa.Table],
+    ) -> MaterializeResult:
+        compacted_table = pa.concat_tables(compacted_tables)
+        if input.compacted_file_content_type in DELIMITED_TEXT_CONTENT_TYPES:
+            # TODO (rkenmi): Investigate if we still need to convert this table to pandas DataFrame
+            # TODO (pdames): compare performance to pandas-native materialize path
+            df = compacted_table.to_pandas(split_blocks=True, self_destruct=True)
+            compacted_table = df
+        delta, stage_delta_time = timed_invocation(
+            input.deltacat_storage.stage_delta,
+            compacted_table,
+            input.write_to_partition,
+            max_records_per_entry=input.max_records_per_output_file,
+            content_type=input.compacted_file_content_type,
+            s3_table_writer_kwargs=input.s3_table_writer_kwargs,
+            **input.deltacat_storage_kwargs,
+        )
+        compacted_table_size = TABLE_CLASS_TO_SIZE_FUNC[type(compacted_table)](
+            compacted_table
+        )
+        logger.debug(
+            f"Time taken for materialize task"
+            f" to upload {len(compacted_table)} records"
+            f" of size {compacted_table_size} is: {stage_delta_time}s"
+        )
+        manifest = delta.manifest
+        manifest_records = manifest.meta.record_count
+        assert manifest_records == len(compacted_table), (
+            f"Unexpected Error: Materialized delta manifest record count "
+            f"({manifest_records}) does not equal compacted table record count "
+            f"({len(compacted_table)})"
+        )
+        materialize_result = MaterializeResult.of(
+            delta=delta,
+            task_index=hash_bucket_index,
+            # TODO (pdames): Generalize WriteResult to contain in-memory-table-type
+            #  and in-memory-table-bytes instead of tight coupling to paBytes
+            pyarrow_write_result=PyArrowWriteResult.of(
```

Comment:
Should we also capture the Parquet format version somewhere during Parquet writes?

Reply:
parquet format version is currently handled as kwargs. We should already be auditing the kwargs.